auto-value-gson/src/main/java/com/ryanharter/auto/value/gson/AutoValueGsonExtension.java

```diff
@@ -688,6 +711,10 @@ private CodeBlock getDefaultValue(Property prop, FieldSpec field) {
         return CodeBlock.of("$T.valueOf(null)", field.type);
       }
     }
```

Comment:
This change asserts that being an optional type takes precedence over being nullable when determining what default value to assign a property. As a result, a nullable optional property will always default to the appropriate empty/absent value during deserialization when the entry is missing from the JSON object or if it is present with a `null` value.

I believe that this is sensible but note that it is somewhat inconsistent with AutoValue. A nullable optional property in AutoValue will default the property to null if the corresponding builder setter is never called.

Reply:
Also note that this defaults `Optional` of a collection type to `Optional.empty()`/`Optional.absent()` and not to an empty collection, regardless of the `defaultCollectionsToEmpty` compiler option. For example, a property of type `Optional<ImmutableSet<String>>`) defaults to `Optional.empty()` and not `Optional.of(ImmutableSet.of())`.

Reply:
This is the change I'm a bit concerned about, and would like to hear the reasoning for the inconsistency with AutoValue.  My understanding is that this basically negates the `@Nullable` annotation completely, as the value could never actually be null.  In practice, this might not matter, as I'm not sure why one would have a nullable optional field, as they seem mutually exclusive.

If the decision is arbitrary, however, I lean towards consistency with AutoValue (if it doesn't complicate things immensely).  I'd appreciate some other input here.

Re: the optional collections, I think that's probably the wrong approach, but, again, would be open to discussion.  My thinking is that `Optional` basically takes the place of `null` to provide better null safety, and the point of the `defaultCollectionToEmpty` option was to provide empty collections in cases where the JSON field was missing.

Most other places users can simply swap `@Nullable` for `Optional` and functionality stays the same, but this changes functionality, basically making the `defaultCollectionToEmpty` a useless compiler option, since it sounds like it does nothing for Optional field.  

Since that compiler option was basically providing a way for users to bandage broken APIs (we always say that the correct fix is to fix your server if it's returning null when you don't want it), we should make sure the bandage still works if users upgrade to `Optional`.

Reply:
(please disregard my deleted comment - I was brainstorming and inadvertently submitted incoherent gibberish).

For the `@Nullable Optional` case:
To be clear, the inconsistency is with `@AutoValue.Builder` and its behavior when a property is not set. I concur that, for consistency with AV and for uniformity in the treatment of `@Nullable`, the default here should be null.

Reply:
For `Optional` collections, I still think that `empty()` is the correct deserialization default. It actually won't change the functionality when swapping `@Nullable` for `Optional` for a collection-valued field, unless I am misreading the code.

`@Nullable` currently takes precedence over `defaultCollectionsToEmpty`. As a result, av-gson currently provides two different, reasonable ways to handle null for collection-valued properties. When `@Nullable` is present on such a property, `defaultCollectionsToEmpty` is not considered for that property.

As with my previous points about `@Nullable Optional`, this use of `@Nullable` is uniform without regard to the property type. If `@Nullable` is omitted, the user can opt-in for an empty collection via the compiler argument. This behavior is also consistent with `@AutoValue.Builder`, whereby omitting a `@Nullable` property results in null by default, regardless of the corresponding property type.

Moreover, when a user wraps a collection-typed property with `Optional`, they are changing the type of property. Unlike `@Nullable`, which is just an annotation, a type change alters the semantics and syntax of how the value must be handled, as Java doesn't permit treating an `Optional` as a proxy for its wrapped type. Because the indirection is explicit, I can't imagine that a user would expect that whatever is wrapped gets special treatment, when the outermost class has a defined way to represent nullness.

Said another way: If the user wants an empty collection by default, why wrap it with `Optional` at all? I suppose there could be other reasons for making a field `Optional` which are unrelated to gson deserialization.

I'm happy to make the changes either way. However, if you do decide that `Optional<Collection<T>>` should default to empty collection, I suggest also changing the behavior of `@Nullable Collection<T>` so that they are consistent.

Reply:
Hi @rharter - just pinging you to get your feedback on our discussion above. I'd be happy to implement the changes as you see fit. Above was just my 2c.